### ISSUE TITLE
Script appears broken - 401 Unauthorized

### ISSUE THREAD

**Issue by phish0r** [pstadler/ticker.sh#33]
/home/ticker.sh AAPL
jq: error (at <stdin>:1): Cannot iterate over null (null)

appears the script is broken

**Comment by phish0r** [comment]
nevermind, script is not broken.  appears to have something to do with the rare google outage that happened today

**Comment by g4570n** [comment]
I am getting the same error trying to use the script. 

`jq: error (at <stdin>:1): Cannot iterate over null (null)`

Does it work for you or does it keep giving the error?

**Comment by phish0r** [comment]
I am receiving the error message again

**Comment by sfsam** [comment]
I am also getting this error. I think the Yahoo finance API is failing and giving jq bad data.

**Comment by phish0r** [comment]
This appears to be an ongoing issue

**Comment by pstadler** [comment]
Yahoo changed something in their finance API. Tried to debug it, but there’s no quick fix.
Basically, they now expect another query parameter which is available from within a browser session. Will need to dig into it a little deeper.

**Comment by pstadler** [comment]
[Here’s](https://github.com/joshuaulrich/quantmod/pull/383/files) how other people are handling this. Two extra calls to obtain a session and fetch a crumb.

Implementation for us would look something like this: If crumb is unavailable or API returns 401, obtain new session, fetch crumb (store it in temporary file for subsequent requests) and repeat the initial call to fetch quotes.

**Comment by servingbaby** [comment]
Oh!
I also just created a quick pull request, testing yours too.

**Comment by servingbaby** [comment]
> Please test this and let me know whether it works and which OS you're using: #35

I got this error at the moment, though I do have this running with root via /usr/local/sbin/:

"mktemp: too few X's in template 'pstadler-ticker-sh'"

**Comment by pstadler** [comment]
> > Please test this and let me know whether it works and which OS you're using: #35
> 
> I got this error at the moment, though I do have this running with root via /usr/local/sbin/:
> 
> "mktemp: too few X's in template 'pstadler-ticker-sh'"

Linux?

**Comment by servingbaby** [comment]
> > > Please test this and let me know whether it works and which OS you're using: #35
> > 
> > 
> > I got this error at the moment, though I do have this running with root via /usr/local/sbin/:
> > "mktemp: too few X's in template 'pstadler-ticker-sh'"
> 
> Linux?

Yes, I'm currently on Arch :)

**Comment by pstadler** [comment]
Updated PR, tested with macos and arch. This should now work in most environments.

**Comment by servingbaby** [comment]
I can confirm working well on my particular environment. Great work and Thank you! @pstadler

**Comment by pstadler** [comment]
merged into master. please pull the latest version.

**Comment by tomaszg7** [comment]
For me (on Gentoo) the old version started to work now (it was broken a few hours ago). New version however doesn't produce any output. No error message, no quotes, nothing.

**Comment by servingbaby** [comment]
> For me (on Gentoo) the old version started to work now (it was broken a few hours ago). New version however doesn't produce any output. No error message, no quotes, nothing.
> 
> Edit: here's content of my crumb file:
> 
> ```
> <!doctype html public "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
>... ```

I wonder if its giving a 500 cause of the cookie file? What does that look like when it gets generated?

**Comment by tomaszg7** [comment]
Cookie file:
```
# Netscape HTTP Cookie File
# https://curl.se/docs/http-cookies.html
# This file was generated by libcurl! Edit at your own risk.

.yahoo.com	TRUE	/	TRUE	1682020284	GUCS	AQOsjLu3
```

**Comment by servingbaby** [comment]
Yep, I think it might be that, 

I was able to recreate your error with your cookies file, but my file worked okay:

$ curl --silent -b cookies.txt "https://query1.finance.yahoo.com/v1/test/getcrumb"
vs
$ curl --silent -b cookies.txt.old "https://query1.finance.yahoo.com/v1/test/getcrumb"
fk22Yb6tiAY

However, my cookies.txt file has a lot more content in there, for example (removing what might be sensitive):
# Netscape HTTP Cookie File
# https://curl.se/docs/http-cookies.html
# This file was generated by libcurl! Edit at your own risk.

.yahoo.com	TRUE	/	TRUE	0	A1S	d=-----
#HttpOnly_.yahoo.com	TRUE	/	TRUE	1713575851	A3	d=-------
#HttpOnly_.yahoo.com	TRUE	/	TRUE	1713575851	A1	d=-------

I am hoping this helps the dev :)

**Comment by pstadler** [comment]
That‘s mighty odd. Thanks for helping out each other here 👍🏻

**Comment by pstadler** [comment]
The HttpOnly cookies are missing. what‘s your curl version?

**Comment by tomaszg7** [comment]
curl 7.88.1 (x86_64-pc-linux-gnu) libcurl/7.88.1 OpenSSL/1.1.1t zlib/1.2.13 brotli/1.0.9 c-ares/1.19.0 nghttp2/1.51.0
Release-Date: 2023-02-20
Protocols: dict file ftp ftps http https imap imaps mqtt pop3 pop3s rtsp smtp smtps tftp
Features: AsynchDNS brotli HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM SSL threadsafe TLS-SRP UnixSockets

**Comment by tomaszg7** [comment]
Also tried updating `curl` and enabling some extra features, but no change:
```
curl 8.0.1 (x86_64-pc-linux-gnu) libcurl/8.0.1 OpenSSL/1.1.1t zlib/1.2.13 brotli/1.0.9 zstd/1.5.4 c-ares/1.19.0 libidn2/2.3.4 nghttp2/1.51.0
Release-Date: 2023-03-20
Protocols: dict file ftp ftps http https imap imaps mqtt pop3 pop3s rtsp smtp smtps tftp
Features: AsynchDNS brotli HSTS HTTP2 HTTPS-proxy IDN IPv6 Largefile libz NTLM SSL threadsafe TLS-SRP UnixSockets zstd
```
Tried it also on another machine (also a similar Gentoo box, but connected to different ISP) and the result was the same.

**Comment by servingbaby** [comment]
This one is a real head scratcher for me

**Comment by pstadler** [comment]
@tomaszg7 please run this and paste the result:

```sh
curl -v "https://finance.yahoo.com" -H "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8"
```

**Comment by tomaszg7** [comment]
```
*   Trying 87.248.106.204:443...
* Connected to finance.yahoo.com (87.248.106.204) port 443 (#0)
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=California; L=Sunnyvale; O=Oath Holdings Inc.; CN=*.api.fantasysports.yahoo.com
*  start date: Apr 10 00:00:00 2023 GMT
*  expire date: May 31 23:59:59 2023 GMT
*  subjectAltName: host "finance.yahoo.com" matched cert's "*.yahoo.com"
*  issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=DigiCert SHA2 High Assurance Server CA
*  SSL certificate verify ok.
* using HTTP/2
* h2h3 [:method: GET]
* h2h3 [:path: /]
* h2h3 [:scheme: https]
* h2h3 [:authority: finance.yahoo.com]
* h2h3 [user-agent: curl/7.88.1]
* h2h3 [accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8]
* Using Stream ID: 1 (easy handle 0x55e9eda3d050)
> GET / HTTP/2
> Host: finance.yahoo.com
> user-agent: curl/7.88.1
> accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
< HTTP/2 307
< date: Fri, 21 Apr 2023 06:31:42 GMT
< strict-transport-security: max-age=31536000
< server: ATS
< cache-control: no-store
< content-type: text/html; charset=utf-8
< content-language: en
< expect-ct: max-age=31536000, report-uri="http://csp.yahoo.com/beacon/csp?src=yahoocom-expect-ct-report-only"
< x-xss-protection: 1; mode=block
< x-content-type-options: nosniff
< location: https://guce.yahoo.com/consent?brandType=nonEu&gcrumb=BlQKcX8&done=https%3A%2F%2Ffinance.yahoo.com%2F
< set-cookie: GUCS=AQZUCnF_; Max-Age=1800; Domain=.yahoo.com; Path=/; Secure
< content-length: 0
<
* Connection #0 to host finance.yahoo.com left intact
```

**Comment by pstadler** [comment]
that’s it, you’re getting redirected:
`< location: https://guce.yahoo.com/consent?brandType=nonEu&gcrumb=BlQKcX8&done=https%3A%2F%2Ffinance.yahoo.com%2F`

**Comment by servingbaby** [comment]
Very interesting!

Something about a consent page for GDPR consent. I wonder if its a geo-location thing that you might be running too that present that redirect..

Out of my comfort zone here, but their might be a way to inject a header to accept the consent cookie and then continue on with the call.

Would you be able to give this a try? :

`curl -v "https://finance.yahoo.com" -H "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8" -H 'cookie: notice_behavior=expressed,eu; notice_gdpr_prefs=0,1,2:1a8b5228dd7ff0717196863a5d28ce6c;'`

(https://stackoverflow.com/questions/64859826/spring-webclient-how-to-skip-accept-cookie-consent-request)

Additional Ref:
https://stackoverflow.com/questions/64912909/getting-consent-cookies-for-website-login-using-python-requests-cmp-euconsent-v
https://adinserter.pro/faq/gdpr-compliance-cookies-consent

**Comment by pstadler** [comment]
![Image from iOS](https://user-images.githubusercontent.com/600097/233647569-4129c206-92e1-48e2-a2a6-e95627cafdf9.jpg)

**Comment by pstadler** [comment]
people are reporting that the API is working again without crumb. Can you confirm this?

**Comment by tomaszg7** [comment]
Here's the output:
```
$ curl -v "https://finance.yahoo.com" -H "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8" -H 'cookie: notice_behavior=expressed,eu; notice_gdpr_prefs=0,1,2:1a8b5228dd7ff0717196863a5d28ce6c;'
*   Trying 87.248.119.252:443...
* Connected to finance.yahoo.com (87.248.119.252) port 443 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS handshake, Client hello (1):
* [CONN-0-0][CF-SSL] TLSv1.3 (IN), TLS handshake, Server hello (2):
* [CONN-0-0][CF-SSL] TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* [CONN-0-0][CF-SSL] TLSv1.3 (IN), TLS handshake, Certificate (11):
* [CONN-0-0][CF-SSL] TLSv1.3 (IN), TLS handshake, CERT verify (15):
* [CONN-0-0][CF-SSL] TLSv1.3 (IN), TLS handshake, Finished (20):
* [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=California; L=Sunnyvale; O=Oath Holdings Inc.; CN=*.api.fantasysports.yahoo.com
*  start date: Apr 10 00:00:00 2023 GMT
*  expire date: May 31 23:59:59 2023 GMT
*  subjectAltName: host "finance.yahoo.com" matched cert's "*.yahoo.com"
*  issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=DigiCert SHA2 High Assurance Server CA
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* h2h3 [:method: GET]
* h2h3 [:path: /]
* h2h3 [:scheme: https]
* h2h3 [:authority: finance.yahoo.com]
* h2h3 [user-agent: curl/7.87.0]
* h2h3 [accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8]
* h2h3 [cookie: notice_behavior=expressed,eu; notice_gdpr_prefs=0,1,2:1a8b5228dd7ff0717196863a5d28ce6c;]
* Using Stream ID: 1 (easy handle 0x5561bae30080)
> GET / HTTP/2
> Host: finance.yahoo.com
> user-agent: curl/7.87.0
> accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
> cookie: notice_behavior=expressed,eu; notice_gdpr_prefs=0,1,2:1a8b5228dd7ff0717196863a5d28ce6c;
> 
* [CONN-0-0][CF-SSL] TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* [CONN-0-0][CF-SSL] TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
< HTTP/2 307 
< date: Fri, 21 Apr 2023 16:42:59 GMT
< strict-transport-security: max-age=31536000
< server: ATS
< cache-control: no-store
< content-type: text/html; charset=utf-8
< content-language: en
< expect-ct: max-age=31536000, report-uri="http://csp.yahoo.com/beacon/csp?src=yahoocom-expect-ct-report-only"
< x-xss-protection: 1; mode=block
< x-content-type-options: nosniff
< set-cookie: notice_behavior=DELETE; Expires=Fri, 21 Apr 2023 16:32:59 GMT; Max-Age=0; Domain=finance.yahoo.com; Path=; Secure; HttpOnly
< set-cookie: notice_behavior=DELETE; Expires=Fri, 21 Apr 2023 16:32:59 GMT; Max-Age=0; Domain=yahoo.com; Path=; Secure; HttpOnly
< set-cookie: notice_gdpr_prefs=DELETE; Expires=Fri, 21 Apr 2023 16:32:59 GMT; Max-Age=0; Domain=finance.yahoo.com; Path=; Secure; HttpOnly
< set-cookie: notice_gdpr_prefs=DELETE; Expires=Fri, 21 Apr 2023 16:32:59 GMT; Max-Age=0; Domain=yahoo.com; Path=; Secure; HttpOnly
< location: https://guce.yahoo.com/consent?brandType=nonEu&gcrumb=Rxwfl3c&done=https%3A%2F%2Ffinance.yahoo.com%2F
< set-cookie: GUCS=AUccH5d3; Max-Age=1800; Domain=.yahoo.com; Path=/; Secure
< content-length: 0
< 
* Connection #0 to host finance.yahoo.com left intact
```

(It's from another machine with another ISP, but still in the same general area).

I confirm that API is working without the crumb, I reported that a few posts back ;) I just checked and it still works.

**Comment by servingbaby** [comment]
> people are reporting that the API is working again without crumb. Can you confirm this?

I am able to confirm :)

commented out the following to test:
```

CRUMB_FILE="${SESSION_DIR}/crumb.txt"
 curl --silent -b "$COOKIE_FILE" "https://query1.finance.yahoo.com/v1/test/getcrumb" \
 "$CRUMB_FILE"
```

and adjusted to this:

`fetch_quotes () {
  curl --silent -b "$COOKIE_FILE" "$API_ENDPOINT&fields=$fields&symbols=$symbols"
}
`

**Comment by pstadler** [comment]
Happily reverted this change!

> I confirm that API is working without the crumb, I reported that a few posts back ;) I just checked and it still works.
@tomaszg7 I ignored your remark at that time, as I observed the API experiencing intermittent issues for days, if not weeks, before it eventually completely broke down. I thought that these disruptions might have been due to A/B testing or similar activities, and I didn't expect it to recover. It seems the joke's on me this time.

If the issue reoccurs, I would be grateful to reach out to both you and @servingbaby for assistance. Thank you for your collaboration!
